### PR TITLE
Added an environment variable to configure the pod IP cooldown - 2378

### PR DESCRIPTION
**Pull request by arturzeitler**
**What type of PR is this?**
feature

**Which issue does this PR fix**:
2378

**What does this PR do / Why do we need it**:
Fixes an open issue.
Added new environment variable ADDRESS_COOLING_PERIOD in charts/aws-vpc-cni/values.yaml expecting an int as string.
pkg/ipamd/datastore/data_store.go defines two envAddressCoolingPeriod with env var name and defaultAddressCoolingPeriod as 30s time.Duration in const section. Then a new function getCoolingPeriod, returning a time.Duration, either returns the default cooling period of 30s or the custom one, if it is not smaller than 0. Also a new logger is defined to log in case the custom duration is used. There was no existing logger in the required scope. getCoolingPeriod is called in inCoolingPeriod with an AddressInfo receiver. Finally, in data_store_test.go, the two hardcoded instances of time.Sleep(30 * time.Second) are replaced by calls to getCoolingPeriod.

**If an issue # is not available please add repro steps and logs from IPAMD/CNI showing the issue**:

**Testing done on this change**:
Only ran the docker unit-tests. have not yet tested it on a running cluster.

**Automation added to e2e**:
No

**Will this PR introduce any new dependencies?**:
No

**Will this break upgrades or downgrades. Has updating a running cluster been tested?**:
Very unlikely. Did not test it on running cluster.

**Does this change require updates to the CNI daemonset config files to work?**:
No

**Does this PR introduce any user-facing change?**:
A new environment variable is added, so kind of user facing.

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: charts/aws-vpc-cni/values.yaml
Comment (jdn5126) [inline charts/aws-vpc-cni/values.yaml:60]: nit: how about `IP_COOLDOWN_PERIOD`?

File: charts/aws-vpc-cni/values.yaml
Comment (jdn5126) [inline charts/aws-vpc-cni/values.yaml:85]: Please revert all of these unnecessary changes

File: pkg/ipamd/datastore/data_store.go
Comment (jdn5126) [inline pkg/ipamd/datastore/data_store.go:274]: This can be a function of `DataStore`, and it can use `ds.log`

File: pkg/ipamd/datastore/data_store.go
Comment (jdn5126) [inline pkg/ipamd/datastore/data_store.go:291]: nit: Set it during DataStore construction and then you can access a member variable instead of having the overhead of a function call each time

**Review by jdn5126 — COMMENTED** [review]
@arturzeitler thanks for the contribution! You will also need to add a README entry for this new environment variable. We should have very strict wording around this being a "use at your own risk" setting.

I also think we need to determine whether 0 is an acceptable value.

Also, we should parse this in `aws-vpc-cni/main.go` and refuse to start if it is not set to a valid integer

File: charts/aws-vpc-cni/values.yaml
Comment (arturzeitler) [reply charts/aws-vpc-cni/values.yaml:60]: Yeah, that's clearer. Will change it.

File: pkg/ipamd/datastore/data_store.go
Comment (arturzeitler) [reply pkg/ipamd/datastore/data_store.go:291]: I'm not sure about this one. If addressCoolingPeriod is set as a member variable of Datastore during NewDataStore, then it would not be available while calling (addr AddressInfo) inCoolingPeriod. Alternatively, addressCoolingPeriod could be a normal variable in the package scope.

File: README.md
Comment (jdn5126) [inline README.md:672]: nit: this will become `v1.13.1+` now

File: README.md
Comment (jdn5126) [inline README.md:678]: nit: I would specify that 0 is not a supported value

File: charts/aws-vpc-cni/values.yaml
Comment (jdn5126) [inline charts/aws-vpc-cni/values.yaml:60]: No entry is needed here, as this is not a mandatory environment variable. Let's revert all changes in this file

File: cmd/aws-vpc-cni/main.go
Comment (jdn5126) [inline cmd/aws-vpc-cni/main.go:104]: nit: let's use a default value as well, to follow the others

File: cmd/aws-vpc-cni/main.go
Comment (jdn5126) [inline cmd/aws-vpc-cni/main.go:364]: nit: let's switch the order of these checks and first validate that it is an integer before doing integer compare

File: pkg/ipamd/datastore/data_store.go
Comment (jdn5126) [inline pkg/ipamd/datastore/data_store.go:82]: nit: let's add a comment here explaining this value

File: README.md
Comment (jdn5126) [inline README.md:680]: On this wording, I think it should be something like:
```
Specifies the number of seconds an IP address is in cooldown after pod deletion. The cooldown period gives kube-proxy time to update node iptables rules when the IP was registered as a valid endpoint, such as for a service. Modify this value with caution, as kube-proxy update time scales with the number of nodes and services.
```

**Review by jdn5126 — CHANGES_REQUESTED** [review]
Changes overall are looking good to me. I just have some nits

**Comment by jdn5126** [comment]
@arturzeitler it looks like there are some conflicts here that need to be resolved, then this can build and run tests

File: cmd/aws-vpc-cni/main.go
Comment (jdn5126) [inline cmd/aws-vpc-cni/main.go:103]: looks like this was added back unintentionally after merge conflict resolution

File: pkg/ipamd/datastore/data_store.go
Comment (jdn5126) [inline pkg/ipamd/datastore/data_store.go:40]: nit: can we update this comment to: "envIPCoolingPeriod controls the time that a deleted pod IP address is held before being eligible to be assigned to a new pod. This holding time gives components that depended on the deleted IP a chance to react to the update."

File: pkg/ipamd/datastore/data_store_test.go
Comment (jdn5126) [inline pkg/ipamd/datastore/data_store_test.go:928]: Can we add a test case where we actually set `IP_COOLING_PERIOD`?

File: pkg/ipamd/datastore/data_store.go
Comment (jdn5126) [inline pkg/ipamd/datastore/data_store.go:275]: nit: why not call the `GetIntAsStringEnvVar` function that was added?

File: cmd/aws-vpc-cni/main.go
Comment (jdn5126) [inline cmd/aws-vpc-cni/main.go:104]: This should be `IP_COOLDOWN_PERIOD`, as that's what the README calls it. Let's also use `envIPCooldownPeriod`

File: cmd/aws-vpc-cni/main.go
Comment (jdn5126) [inline cmd/aws-vpc-cni/main.go:361]: `IP_COOLDOWN_PERIOD`

File: cmd/aws-vpc-cni/main.go
Comment (jdn5126) [inline cmd/aws-vpc-cni/main.go:365]: `IP_COOLDOWN_PERIOD`

**Review by jdn5126 — CHANGES_REQUESTED** [review]
Sorry for the delay @arturzeitler . A few more comments and then I think this is good to go

File: pkg/ipamd/datastore/data_store_test.go
Comment (arturzeitler) [reply pkg/ipamd/datastore/data_store_test.go:928]: yes, it would help though if you could point me to a similar test case.

**Comment by jdn5126** [comment]
@arturzeitler are you still able to drive this? I am good to merge this after we rebase, fix a few more nits, and pass a run of integration tests

**Comment by jdn5126** [comment]
Closing as https://github.com/aws/amazon-vpc-cni-k8s/pull/2492 is taking this over

**Pull request closed without merging**
